stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
+ public:
+  static_assert(
+      is_plain_type<T>::value,
+      "The template for this var is an"
+      " expression but a var_value's inner type must be assignable such as"
+      " a double, Eigen::Matrix, or Eigen::Array");
+
+  /**
+   * `PlainObject` represents a user constructible type such as Matrix or Array
+   */
+  using PlainObject = std::decay_t<plain_type_t<T>>;
+  using Scalar = PlainObject;  // The underlying type for this class
+  using value_type = Scalar;   // The underlying type for this class
+  using eigen_scalar = value_type_t<PlainObject>;  // A floating point type
+  /**
+   * Maps for adj_ and val_
+   */
+  using eigen_map = Eigen::Map<PlainObject, Eigen::Aligned8>;
+  using vari_type = vari_value<T, require_eigen_dense_base_t<T>>;
+  eigen_scalar* val_mem_;  // Pointer to memory allocated on the stack for val_
+  eigen_scalar* adj_mem_;  // Pointer to memory allocated on the stack for adj_
+  /**
+   * Number of rows known at compile time
+   */
+  static constexpr Eigen::Index RowsAtCompileTime = Scalar::RowsAtCompileTime;
+  /**
+   * Number of columns known at compile time
+   */
+  static constexpr Eigen::Index ColsAtCompileTime = Scalar::ColsAtCompileTime;
+
+  /**
+   * The value of this variable.
+   */
+  const eigen_map val_;
+
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  eigen_map adj_;
+
+  /**
+   * Construct a dense Eigen variable implementation from a value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  explicit vari_value(S&& x)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(eigen_map(val_mem_, x.rows(), x.cols()) = x),
+        adj_(eigen_map(adj_mem_, x.rows(), x.cols()).setZero()) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct an dense Eigen variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
```

Comment:
This paragraph seems outdated considering the one above it.

Reply:
I think both paragraphs can be combined to remove duplicate info